core/templates/pages/exploration-player-page/services/command-executor.service.spec.ts

```diff
@@ -0,0 +1,379 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the command executor service
+ */
+
+import { TestBed } from '@angular/core/testing';
+import { CommandExecutorService } from
+  'pages/exploration-player-page/services/command-executor.service';
+import { WindowRef } from 'services/contextual/window-ref.service.ts';
+describe('Command executor service', () => {
+  let ces: CommandExecutorService, wrf: WindowRef;
+  var spy;
+  beforeEach((done) => {
+    TestBed.configureTestingModule({
+      providers: [CommandExecutorService, WindowRef]
+    });
+    ces = TestBed.get(CommandExecutorService);
+    wrf = TestBed.get(WindowRef);
+    setupWindowRef(wrf);
+    spy = spyOn(ces.windowWrapperMessageService, 'addEventListener');
+    ces.getOuterFrameEvents(wrf);
+    done();
+  });
+
+  var continueBoolean = false;
+  var addBoolean = false;
+  var deleteBoolean = false;
+  var mcBoolean = false;
+  var secondaryContinueBoolean = false;
+  var setupWindowRef = function(windowRef: WindowRef) {
+    var suite =
+        wrf.nativeWindow.document.getElementsByTagName('TESTING_SUITE')[0];
+    if (suite) {
+      suite.remove();
+    }
+    continueBoolean = false;
+    addBoolean = false;
+    deleteBoolean = false;
+    mcBoolean = false;
+    secondaryContinueBoolean = false;
+    var testPage = windowRef.nativeWindow.document.createElement(
+      'TESTING_SUITE');
+    var continueButton = windowRef.nativeWindow.document.createElement(
+      'BUTTON');
+    continueButton.classList.add(
+      'oppia-learner-confirm-button');
+    continueButton.onclick = function() {
+      continueBoolean = true;
+    };
+    var secondaryContinueButton =
+      windowRef.nativeWindow.document.createElement('BUTTON');
```

Comment:
This is not the right place for all these tests. DOM tests should be done using the Protractor e2e framework. See also the existing test pages in assets/scripts, and the embedding.js Protractor test, for some ideas that you might be able to base things on.

Karma tests should be used mainly for the logic between controllers, directives, etc. -- make sure the correct functions get called when some event is received/output by the CommandExecutorService, etc. But in general they should not have a DOM component, and all this creation of HTML elements is a bit of an antipattern for the Karma tests. (Note that we've also experienced quite a bit of flakiness in the past when working with DOM elements in Karma tests.)